docs/weave-tutorial/07-handlers.md

```diff
@@ -0,0 +1,675 @@
+replayed. Even the slightest deviation will cause the merkle root
+of the store at the end of the block to differ with other nodes,
+and thus kick the deviating nodes out of consensus.
+(Note that _Check_ may actually vary between nodes without breaking
+consensus rules, although we generally keep this deterministic as well).
+
+**This is a very powerful concept and means that when modifying a given state,
+users must not worry about any concurrent access or writing collision
+since by definition, any write access is guaranteed to occur sequentially and
+in the same order on each node**
+
+## Writing a Handler
+
+We usually can write a separate handler for each message type,
+although you can register multiple messages with the same
+handler if you reuse most of the code. Let's focus on the
+simplest cases, and the handlers for creating a Blog and
+adding a Post to an existing blog.
+
+Note that we can generally assume that _Handlers_ are wrapped
+by a `Savepoint Decorator`, and that if _Deliver_ returns an error after updating some
+objects, those update will be discarded. This means you can
+treat _Handlers_ as atomic actions, all or none, and not worry
+too much about cleaning up partially finished state changes
+if a later portion fails.
+
+## Validation
+
+Remember that we have to fulfill both _Check_ and _Deliver_ methods,
+and they share most of the same validation logic. A typical
+approach is to define a _validate_ method that parses the
+proper message out of the transaction, verify all authorization
+preconditions are fulfilled by the transaction, and possibly
+check the current state of the blockchain to see if the action
+is allowed. If the _validate_ method doesn't return an error,
+then _Check_ will return the expected cost of the transaction,
+while _Deliver_ will actually perform the action and update
+the blockchain state accordingly.
+
+## Examples
+
+```go
+// CreateBlogHandler will handle CreateBlogMsg
+type CreateBlogHandler struct {
+    auth x.Authenticator
+    b    *BlogBucket
+}
+
+var _ weave.Handler = CreateBlogHandler{}
+
+// NewCreateBlogHandler creates a blog message handler
+func NewCreateBlogHandler(auth x.Authenticator) weave.Handler {
+    return CreateBlogHandler{
+        auth: auth,
+        b:    NewBlogBucket(),
+    }
+}
+```
+
+First we define what the `CreateBlogHandler` which components it will be operating on. `x.Authenticator` will handle the authentication and `BlogBucket` will be the gateway to the blog data. `CreateBlogHandler` is just a struct to wrap the components that is required to make handler logic work, such as **authentication** and **data access**. Authentication is handled by `x.Authenticator` and blog data access is by `BlogBucket` that we defined in previous chapter. `var _ weave.Handler = CreateBlogHandler{}` is just a helper to ensure `CreateBlogHandler` is a `weave.Handler`. `NewCreateBlogHandler` wraps the components and returns the struct.
```

Comment:
```suggestion
First we define what the `CreateBlogHandler` which components it will be operating on. `x.Authenticator` will handle the authentication and `BlogBucket` will be the gateway to the blog data. `CreateBlogHandler` is just a struct to wrap the components that are required to make handler logic work, such as **authentication** and **data access**. Authentication is handled by `x.Authenticator` and blog data access is by `BlogBucket` that we defined in the previous chapter. `var _ weave.Handler = CreateBlogHandler{}` is just a helper to ensure `CreateBlogHandler` is a `weave.Handler`. `NewCreateBlogHandler` wraps the components and returns the struct.
```